junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncher.java

```diff
@@ -132,6 +155,9 @@ private Root discoverRoot(LauncherDiscoveryRequest discoveryRequest, String phas
 				() -> String.format(
 					"The discover() method for TestEngine with ID '%s' must return a non-null root TestDescriptor.",
 					testEngine.getId()));
+			Preconditions.condition(isAcyclic(engineRoot),
+				() -> String.format("The discover() method for TestEngine with ID '%s' returned a cyclic graph.",
+					testEngine.getId()));
```

Comment:
I think we should extract these checks into a separate `EngineDiscoveryResultValidator` class or sth. similar.